doc/generated/man/irqloop.1
'\" t
.\"     Title: irqloop
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 2008/04/19
.\"    Manual: Xenomai Manual
.\"    Source: Xenomai 2.6.5
.\"  Language: English
.\"
.TH "IRQLOOP" "1" "2008/04/19" "Xenomai 2\&.6\&.5" "Xenomai Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
irqloop \- Xenomai IRQ benchmark, target program
.SH "SYNOPSIS"
.sp
\fBirqloop\fR [ options ]
.SH "DESCRIPTION"
.sp
\fBirqloop\fR is part of the Xenomai test suite and benchmarks IRQs\&. It is the program running on the target machine which is running a suitable Xenomai enabled kernel with the respective module (xeno_posix)\&. The additional tool \fBirqbench(1)\fR is running on a plain Linux host, connected via RS232 or Laplink with the target\&.
.SH "OPTIONS"
.sp
\fBirqloop\fR accepts the following options:
.PP
\fB\-D <testing_device_no>\fR
.RS 4
number of testing device, default=0
.RE
.PP
\fB\-t <test_mode>\fR
.RS 4
0=user task (default) 1=kernel task 2=IRQ handler 3=hard\-IRQ handler
.RE
.PP
\fB\-P <priority>\fR
.RS 4
task priority (test mode 0 and 1 only)
.RE
.PP
\fB\-o <port_type>\fR
.RS 4
0=serial (default), 1=parallel
.RE
.PP
\fB\-a <port_io_address>\fR
.RS 4
default=0x3f8/0x378
.RE
.PP
\fB\-i <port_irq>\fR
.RS 4
default=4/7
.RE
.SH "SEE ALSO"
.sp
\fB/usr/share/doc/xenomai\-doc/txt/irqloop\&.txt, \fR\fBirqbench(1)\fR\&.
.SH "AUTHOR"
.sp
\fBirqloop\fR was written by Jan Kiszka\&. This man page was written by Roland Stigge\&.


ChangeLog
2009-03-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vxworks/semLib.c (sem_destroy_internal) : Add missing
	rescheduling. Fix sem_read_proc().

2009-03-19  Philippe Gerum  <rpm@xenomai.org>

	* include/asm-generic/wrappers.h, src/nucleus/shadow.c: Fix
	generic bits for 2.6.29.

2009-03-18  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/pipe.c: Fix streaming mode.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/nucleus/shadow.c: Trace syscall exits.

	* patching file include/asm-generic/syscall.h,
	ksrc/nucleus/shadow.c: Add xn_sys_current_info.

	* src/rtdk/rt_print.c: suspend/resume printer thread.

	* ksrc/skins/native/syscall.c, src/skins/native/task.c: Support
	T_WARNSW on task creation.

	* src/skins/native/task.c: Avoid lazy non-TLS rt_task_self.

	* src/skins/native/task.c: Silent migration before self-deletion.

	* src/skins, ksrc/nucleus/shadow.c: Rework initial Linux
	schedparam setting.

	* ksrc/nucleus, include/asm-generic, ksrc/arch/generic: Exclude
	CPUs from Xenomai use.

	* ksrc/nucleus/shadow.c: Cleanup taskexit hook.

	* ksrc/nucleus/shadow.c: Suppress switch warning upon signal
	receipt when the process is being debugged.

	* ksrc/nucleus/sched.c: Allow to change the watchdog timeout via a
	module parameter.

	* ksrc/arch/x86/usercopy_{32, 64}.c: Fix miscompilation of
	__do_strncpy_from_user.

	* src/rtdk, include/rtdk.h: Add assert_context helpers.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/skins/posix/syscall.c: Do not return -EFAULT on zero-length
	names. Instead, return -EINVAL when trying to create objects with
	empty names.

2009-03-13  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/mx.c: Streamline mutex support.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/task.c, src/skins/vrtx/task.c: Fix task creation.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/shadow.c, ksrc/nucleus/synch.c: Introduce debug
	option to detect mutexes held in relaxed sections.

2009-03-07  Philippe Gerum  <rpm@xenomai.org>

	* src/skins/posix/shm.c (__wrap_munmap): Forward to glibc upon EBADF.

	* ksrc/nucleus/Config.in, ksrc/arch/{powerpc, x86}/Config.in: Add
	missing SYS_STACKPOOLSZ switch.

2009-02-26  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/powerpc/patches: Update Adeos support to
	2.6.24.6-powerpc-DENX-2.4-03 (compatible with MPC5121-ADS
	hardware, see ksrc/arch/powerpc/patches/README).

2009-02-25  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/nmi.c: Rework NMI watchdog support.
	
2008-02-24  Stephane Fillod  <stephane.fillod@thomson.net>

	* src/skins/posix/shm.c (__wrap_mmap64): Make real call depend on
	configure detection (HAVE_MMAP64).
	
2009-02-24  Yi Li <liyi.dev@gmail.com>

	* configure.in: Remove -mfdpic switch. Set by default by
	bfin-linux-uclibc-gcc, and may cause error with bfin-uclinux-gcc.

2009-02-24  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/{arm, blackfin, powerpc}/hal.c (rthal_timer_request):
	Port x86 fix of race window.

2009-02-24  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/hal-common.c (rthal_timer_request): Fix race.

	* ksrc/nucleus/shadow.c (gatekeeper_thread): Do not refer to
	uninit sched slots.

2009-02-23  Paul Corner <paul_c@tuxcnc.org>

	* debian/prepare-patch.sh: Fix patch generation rules.

2009-02-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.19-x86-2.2-06,
	2.6.28.7-x86-2.2-06.

2009-02-17  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-05.

	* ksrc/arch/powerpc/patches: Upgrade to
	2.6.28.5-powerpc-DENX-2.4-04.

2009-02-13  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* src/skins/posix/shm.c (__wrap_mmap, __wrap_mmap64): Map only
	part of the underlying nucleus heap.

	* ksrc/nucleus/heap.c (xnheap_mmap): Allow partial mapping in
	user-space of a nucleus heap.

	* ksrc/nucleus/sched.c, include/nucleus/sched.h: Introduce
	sched->cpu field member.

2009-02-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pod.c (xnpod_start_thread): Run start hook for
	user-space threads as well.

2009-02-11  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/posix/shm.c (ftruncate): Remove GFP_USER hint.

2009-02-07  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/blackfin, include/asm-blackfin: Prepare for 2.6.28.

	* ksrc/arch/blackfin/patches: Upgrade to 2.6.28-1.9-00 (GIT).

	* ksrc/nucleus, include/nucleus: Introduce host tick deferral.

2009-02-02  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/syscall.c (rt_queue_write): Fix memory leak.

2009-02-01  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/heap, include/asm-generic/system.h, ksrc/skins: Fix
	auto-cleanup for shared heaps.

2009-01-31  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/drivers/testing/switchtest.c,
	  src/testsuite/switchtest/switchtest.c: Add stress mode to
	switchtest.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/ia64, include/asm-ia64: Itanium port discontinued.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-04,
	2.6.28.2-x86-2.2-05.

2009-01-27  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* include/asm-x86/bits/pod_{32, 64}.h (xnarch_leave_root): Fix FPU
	backup clobbering.
	
2009-01-27  Philippe Gerum  <rpm@eclipse.xenomai.org>

	* ksrc/nucleus, include/nucleus, ksrc/skins/posix, include/posix,
	sim, include/asm-sim: Introduce POSIX sporadic scheduling (PSS)
	class.

2009-01-21  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/arch/arm/patches: Upgrade to 2.6.28-arm-1.2-00.
	
2009-01-19  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pipe.c, ksrc/skins/native/pipe.c,
	ksrc/skins/rtai/fifo.c: Refactor callback management, fixing a
	severe latency issue, and a potential memory leak in both
	disconnect/release handlers.

	* ksrc/nucleus/pipe.c (xnpipe_disconnect, xnpipe_release):

src/testsuite/irqbench/irqloop.c
/*
 * Copyright (C) 2006 Jan Kiszka <jan.kiszka@web.de>.
 *
 * Xenomai is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Xenomai is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Xenomai; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <string.h>
#include <pthread.h>
#include <unistd.h>
#include <limits.h>
#include <sys/mman.h>
#include <rtdm/rttesting.h>

static int benchdev;
static int terminate;

static void *irq_thread(void *arg)
{
	struct sched_param param = { .sched_priority = (long)arg };

	pthread_setschedparam(pthread_self(), SCHED_FIFO, &param);

	while (1) {
		if (ioctl(benchdev, RTTST_RTIOC_IRQBENCH_WAIT_IRQ) ||
		    ioctl(benchdev, RTTST_RTIOC_IRQBENCH_REPLY_IRQ))
			break;
	}

	param.sched_priority = 0;
	pthread_setschedparam(pthread_self(), SCHED_OTHER, &param);

	return NULL;
}

static void sighand(int sig)
{
	terminate = 1;
}

int main(int argc, char *argv[])
{
	const char *mode_name[] = {
		[RTTST_IRQBENCH_USER_TASK]	= "user-space task",
		[RTTST_IRQBENCH_KERNEL_TASK]	= "kernel-space task",
		[RTTST_IRQBENCH_HANDLER]	= "IRQ handler",
		[RTTST_IRQBENCH_HARD_IRQ]	= "hard-IRQ handler",
	};
	const char *port_type_name[] = {
		[RTTST_IRQBENCH_SERPORT]	= "serial",
		[RTTST_IRQBENCH_PARPORT]	= "parallel",
	};
	char devname[RTDM_MAX_DEVNAME_LEN];
	int benchdev_no = 0;
	struct rttst_irqbench_config config = {
		.mode			= RTTST_IRQBENCH_USER_TASK,
		.priority		= sched_get_priority_max(SCHED_FIFO),
		.calibration_loops	= 0,
		.port_type		= RTTST_IRQBENCH_SERPORT,
		.port_ioaddr		= 0x3f8,
		.port_irq		= 4
	};
	struct rttst_irqbench_stats stats;
	unsigned long long last_received = 0;
	pthread_t thr;
	int ioaddr_set = 0;
	int irq_set = 0;
	int c;
	int timeout = 10;

	while ((c = getopt(argc, argv, "D:t:P:o:a:i:")) != EOF)
		switch (c) {
		case 'D':
			benchdev_no = atoi(optarg);
			break;

		case 't':
			config.mode = atoi(optarg);
			break;

		case 'P':
			config.priority = atoi(optarg);
			break;

		case 'o':
			config.port_type = atoi(optarg);
			break;

		case 'a':
			config.port_ioaddr = strtol(optarg, NULL, 0);
			ioaddr_set = 1;
			break;

		case 'i':
			config.port_irq = strtol(optarg, NULL, 0);
			irq_set = 1;
			break;

		default:
			fprintf(stderr, "usage: irqloop [options]\n"
				"  [-D <testing_device_no>] # number of testing device, default=0\n"
				"  [-t <test_mode>]         # 0=user task (default), 1=kernel task,\n"
				"                           # 2=IRQ handler, 3=hard-IRQ handler\n"
				"  [-P <priority>]          # task priority (test mode 0 and 1 only)\n"
				"  [-o <port_type>]         # 0=serial (default), 1=parallel\n"
				"  [-a <port_io_address>]   # default=0x3f8/0x378\n"
				"  [-i <port_irq>]          # default=4/7\n");
			exit(2);
		}

	/* set defaults for parallel port */
	if (config.port_type == 1) {
		if (!ioaddr_set)
			config.port_ioaddr = 0x378;
		if (!irq_set)
			config.port_irq = 0x7;
	}

	signal(SIGINT, sighand);
	signal(SIGTERM, sighand);
	signal(SIGHUP, sighand);

	mlockall(MCL_CURRENT|MCL_FUTURE);

	snprintf(devname, RTDM_MAX_DEVNAME_LEN, "/dev/rttest-irqbench%d", benchdev_no);
	benchdev = open(devname, O_RDWR);
	if (benchdev < 0) {
		perror("irqloop: failed to open benchmark device");
		fprintf(stderr, "(modprobe xeno_irqbench?)\n");
		return 1;
	}

	if (config.mode == RTTST_IRQBENCH_USER_TASK) {
		pthread_attr_t attr;

		pthread_attr_init(&attr);
		pthread_attr_setstacksize(&attr, PTHREAD_STACK_MIN);

		pthread_create(&thr, &attr, irq_thread, (void *)(long)config.priority);
	}

	if (ioctl(benchdev, RTTST_RTIOC_IRQBENCH_START, &config)) {
		perror("irqloop: error starting test");
		goto cleanup;
	}

	printf("Test mode:    %s\n"
	       "Port type:    %s\n"
	       "Port address: 0x%lx\n"
	       "Port IRQ:     %d\n\n\n\n",
	       mode_name[config.mode], port_type_name[config.port_type],
	       config.port_ioaddr, config.port_irq);

	while (!terminate) {
		if (ioctl(benchdev, RTTST_RTIOC_IRQBENCH_GET_STATS,
			  &stats) < 0) {
			perror("irqloop: error reading stats");
			break;
		}

		if ((last_received > 0) &&
		    (stats.irqs_received == last_received)) {
			if (--timeout == 0)
				break; /* timed out */
		} else
			timeout = 10;
		last_received = stats.irqs_received;

		printf("\033[2AReceived IRQs:     %lld\n"
		       "Acknowledged IRQs: %lld\n",
		       stats.irqs_received, stats.irqs_acknowledged);
		usleep(250000);
	}

	ioctl(benchdev, RTTST_RTIOC_IRQBENCH_STOP);

cleanup:
	close(benchdev);
	if (config.mode == RTTST_IRQBENCH_USER_TASK) {
		pthread_cancel(thr);
		pthread_join(thr, NULL);
	}
	return 0;
}
